src/compiler/crystal/syntax/parser.cr

```diff
@@ -5662,11 +5662,14 @@ module Crystal
       check :","
 
       next_token_skip_space_or_newline
-      check :INSTANCE_VAR
+      check(%i(INSTANCE_VAR NUMBER))
 
       ivar_location = @token.location
-      instance_var = InstanceVar.new(@token.value.to_s).at(ivar_location)
-
+      instance_var = if @token.type == :INSTANCE_VAR
+                       InstanceVar.new(@token.value.to_s).at(ivar_location)
+                     else
+                       NumberLiteral.new(@token.value.to_s, @token.number_kind)
```

Comment:
You can do `if @token.type == :INSTANCE_VAR ... elsif @token.type == :NUMBER ... else raise "expecting instance variable or number literal, not ...".

And here you can also check that the number is not a float and give that as a parse error, instead of a semantic error later on.

Reply:
Yep, I was in doubt where to throw this error, in the parser or later, mainly when I saw that the number token is for integers and floats, TL;DR; I agree, better let it be a parser error.